### DISCUSSION TITLE
Trying to compile...something is missing...

### DISCUSSION THREAD

**Discussion by n3gwg** [ipxe/ipxe#961] [General]
GOAL: I wanted to have a version of ipxe.efi that included the full compliment of shell commands (like ping, which is missing from the version on the website). Thus, I attempted to build and compile from source.

I believe I installed all the requisites as mentioned in the documentation on the website, then followed the README. I even made sure to do a make clean first to assure the compilation environment was ready to go.

~/source/ipxe $ cat README 
iPXE README File

Quick start guide:

   cd src
   make

For any more detailed instructions, see http://ipxe.org

~/source/ipxe $ cd src

~/source/ipxe/src $ make clean
rm -f -r bin/*.* bin-*/*.* bin/.certificate.* bin-*/.certificate.* bin/.certificates.* bin-*/.certificates.* bin/.private_key.* bin-*/.private_key.* bin/errors bin-*/errors bin/NIC bin-*/NIC ./util/zbin ./util/zbin ./util/elf2efi32 ./util/elf2efi32 ./util/elf2efi64 ./util/elf2efi64 ./util/efirom ./util/efirom ./util/efifatbin ./util/efifatbin ./util/einfo ./util/einfo TAGS TAGS bin/symtab bin-*/symtab

~/source/ipxe/src $ make
  [DEPS] crypto/privkey.c
  [DEPS] crypto/certstore.c

[...clipped a lot of text...]

  [BUILD] bin/ath9k.ids.o
  [AR] bin/blib.a
ar: creating bin/blib.a
  [HOSTCC] util/zbin
  [VERSION] bin/version.ipxe.dsk.o
  [LD] bin/ipxe.dsk.tmp
ld: warning: patch_cf.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
ld: warning: bin/ipxe.dsk.tmp has a LOAD segment with RWX permissions
ld: warning: bin/ipxe.dsk.tmp has a LOAD segment with RWX permissions
ld: warning: bin/ipxe.dsk.tmp has a LOAD segment with RWX permissions
ld: warning: bin/ipxe.dsk.tmp has a LOAD segment with RWX permissions
  [BIN] bin/ipxe.dsk.bin
  [ZINFO] bin/ipxe.dsk.zinfo
  [ZBIN] bin/ipxe.dsk.zbin
  [FINISH] bin/ipxe.dsk
  [VERSION] bin/version.ipxe.lkrn.o
  [LD] bin/ipxe.lkrn.tmp
ld: warning: patch_cf.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
ld: warning: bin/ipxe.lkrn.tmp has a LOAD segment with RWX permissions
ld: warning: bin/ipxe.lkrn.tmp has a LOAD segment with RWX permissions
ld: warning: bin/ipxe.lkrn.tmp has a LOAD segment with RWX permissions
ld: warning: bin/ipxe.lkrn.tmp has a LOAD segment with RWX permissions
ld: warning: bin/ipxe.lkrn.tmp has a LOAD segment with RWX permissions
  [BIN] bin/ipxe.lkrn.bin
  [ZINFO] bin/ipxe.lkrn.zinfo
  [ZBIN] bin/ipxe.lkrn.zbin
  [FINISH] bin/ipxe.lkrn
  [GENFSIMG] bin/ipxe.iso
util/genfsimg: could not find isolinux.bin
make: *** [arch/x86/Makefile.pcbios:61: bin/ipxe.iso] Error 1
rm bin/ipxe.dsk.zinfo bin/ipxe.lkrn.bin bin/ipxe.lkrn.zinfo bin/ipxe.lkrn.zbin bin/ipxe.dsk.zbin bin/version.ipxe.dsk.o bin/ipxe.dsk.bin bin/version.ipxe.lkrn.o

~/source/ipxe/src $ sudo dpkg --list | grep -Ei 'syslinux|extlinux'
ii  extlinux                                                    3:6.04~git20190206.bf6db5b4+dfsg1-3+b1 amd64        collection of bootloaders (Linux ext2/ext3/ext4, btrfs, and xfs bootloader)
ii  syslinux                                                    3:6.04~git20190206.bf6db5b4+dfsg1-3+b1 amd64        collection of bootloaders (DOS FAT and NTFS bootloader)
ii  syslinux-common                                             3:6.04~git20190206.bf6db5b4+dfsg1-3    all          collection of bootloaders (common)

~/source/ipxe/src $ sudo find / -name 'islinux.bin' -exec ls -lah {} \; 2>/dev/null

~/source/ipxe/src $ 

Clearly, isolinux.bin is not there! However, I have a fully and properly installed version of syslinux (I am using Kali Linux as my distribution). Any ideas on why it is missing? Was it removed from syslinux at some point?

Thanks in advance!

Stuart

PS - some of the above text is "lined through", I have no idea how or why that became occurring.

**Comment by n3gwg** [comment]
This was apparently moved from being an issue to the discussion area. I am not sure why as if a compile fails because a makefile is unable to trap an error and go on to a phase of the compile that does not require what the former phase needed, that to me is a bug and an issue.

**Comment by NiKiZe** [comment]
TL;DR: try building using `make bin-x86_64-efi/ipxe.efi`

## why moved to discussions
Moved to discussion as this is a case of not having the required dependencies (listed at https://ipxe.org/download#source_code) for the chosen build.

## what is going on?
`make` without specifying target builds the default i386 pcbios variants, including the iso, the iso requires `isolinux.bin` (you wrote `islinux.bin` in your search.
However if you want ipxe.efi, then what you probably want is `make bin-x86_64-efi/ipxe.efi`

Other targets can be found at https://ipxe.org/appnote/buildtargets

## still want the pcbios iso
If you want a a iso with pcbios support, then `isolinux.bin` is required, and you will have to check if it exists on your system, and which package provides it.

## But this is still an error with the build...
isolinux is listed as a requirement for building isos,
Following the instructions for build requirements or not building pcbios ISOs and you would not get any error.

With that said, @mcb30 now days we might be better of not defaulting to building the ISO, or not even defaulting to the pcbios bin build at all and instead just show a quick help on which targets you might want to build.

**Reply by n3gwg** [reply]
Let me first start off by saying that I appreciate most greatly the totality of your reply. Moreover, I am generally indifferent into which category this colloquy goes, just I really did sincerely feel that there is a bug or failure in the makefile code, which I will address further on in my response.

> TL;DR: try building using `make bin-x86_64-efi/ipxe.efi`
> ## why moved to discussions
> 
> Moved to discussion as this is a case of not having the required dependencies (listed at https://ipxe.org/download#source_code) for the chosen build.

Worthy of notation is that I did think this at first. However, succeeding a review of the requisite packages and having verified there instantiation within my environment (whilst still taking notice of the absence of isolinux.bin), I interpreted the situation to be a defect in the instructions at minimum.

You are correct I got the search wrong, but even redoing the find with the proper filename I did not find it upon my system despite verifying that syslinux and extlinux were properly installed thereupon.

> ## what is going on?
> 
> `make` without specifying target builds the default i386 pcbios variants, including the iso, the iso requires `isolinux.bin` (you wrote `islinux.bin` in your search. However if you want ipxe.efi, then what you probably want is `make bin-x86_64-efi/ipxe.efi`
> 
> Other targets can be found at https://ipxe.org/appnote/buildtargets

What I wanted to do was build the efi, non-efi, ISO and USB versions, so it seemed doing the full build was the best way to achieve that. I have no 386 machines and had never an intention to support any 386 based code.

> ## still want the pcbios iso
> 
> If you want a a iso with pcbios support, then `isolinux.bin` is required, and you will have to check if it exists on your system, and which package provides it.

I understand and do not dispute this, I am just pointing out that installing all the packages as indicated as being required did not get this critical file onto my system. That obviously left me rather bewildered.

> ## But this is still an error with the build...
> 
> isolinux is listed as a requirement for building isos, Following the instructions for build requirements or not building pcbios ISOs and you would not get any error.
> 
> With that said, @mcb30 now days we might be better of not defaulting to building the ISO, or not even defaulting to the pcbios bin build at all and instead just show a quick help on which targets you might want to build.

I would agree with that.

Let me further indicate that I ended up realizing that I did not need to build iPXE at all. I am using netboot.xyz and wanted to have a preemptive menu, which I got working by creating a menu.ipxe file on my TFTP server (I use pfSense for routing and network functions, thus DHCP and TFTP service come from pfSense for me). However, even the version of iPXE that was on the website was missing commands like ping and that which were present in the iPXE build that was inclusive to netboot.xyz, so I was a bit confused by that too.

Thanks again for all your help and I am sorry I did not originally place the request for assistance in the correct place. I felt that if the makefile "just breaks and stops" (even if due to a deficiency on my development environment), it should try to continue to build what all else it can or indicate that no further building can be done due to this deficiency.

One last question I am curious about is that I noticed you can sanboot an ISO file via HTTP, does sanboot for an ISO file also work with HTTPS or strictly HTTP?

Stuart

**Reply by NiKiZe** [reply]
Any pcbios build is i386 compatible.
make on its own does not build efi.
iPXE can not keep track of every distro and it's packages. (In your case they butchered the syslinix package)

Yes you can sanboot over any protocol that supports range requests, but do not be filled in to thinking that you can boot any ISOs from iPXE without issues. (Other issue that is covered plenty of times)

**Reply by n3gwg** [reply]
> Any pcbios build is i386 compatible. make on its own does not build efi. iPXE can not keep track of every distro and it's packages. (In your case they butchered the syslinix package)

I agree entirely and would think such an endeavor of tracking every distro and its packages would be insane, that is why I meant to suggest that merely saying having syslinux installed in the instructions is not instructive enough. The instructions should substantiate with greater granularity a warning like, "Even if the syslinux package is installed, you must assure isolinux.bin is on your system, as if it is not the build will fail notwithstanding the presence of syslinux."

> 
> Yes you can sanboot over any protocol that supports range requests, but do not be filled in to thinking that you can boot any ISOs from iPXE without issues. (Other issue that is covered plenty of times)

I had read that sanboot was not a perfect answer, but I do appreciate being reminded of this.

Stuart

**Comment by rstms** [comment]
I came here looking to resolve:  ```util/genfsimg: could not find isolinux.bin```
On Debian 12 (bookworm), I added the package ```isolinux``` which provides ```isolinux.bin```
```
mkrueger@ipxe-builder:~/src$ apt-file search /usr/lib/ISOLINUX/isolinux.bin
isolinux: /usr/lib/ISOLINUX/isolinux.bin
```

**Reply by NiKiZe** [reply]
Then I think the last path in this `for` should find it. https://github.com/ipxe/ipxe/blob/7f75d320f6d8ac7ec5185b2145da87f698aec273/src/util/genfsimg#L97 and as it says on the download page, isolinux is a requirement (even if it isn't clear that it is only required to build ISOs)
